pkg/client/rootd/dns/connpool.go

```diff
@@ -0,0 +1,111 @@
+package dns
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+type ConnPool struct {
+	items      map[*dns.Conn]bool
+	newArrival chan *waitingClient
+	finished   chan *dns.Conn
+	clients    clientQueue
+	cancel     context.CancelFunc
+	RemoteAddr string
+}
+
+func NewConnPool(addr string, poolSize int) *ConnPool {
+	cCtx, cCancel := context.WithCancel(context.Background())
+	pool := &ConnPool{
+		items:      make(map[*dns.Conn]bool, poolSize),
+		newArrival: make(chan *waitingClient),
+		finished:   make(chan *dns.Conn),
+		cancel:     cCancel,
+		RemoteAddr: addr,
+	}
+	heap.Init(&pool.clients)
+	for i := 0; i < poolSize; i++ {
+		conn, err := dns.Dial("udp", net.JoinHostPort(addr, "53"))
+		if err != nil {
+			fmt.Printf("failed to create conn: %v", err)
+			os.Exit(1)
+		}
+		pool.items[conn] = false
+	}
+	go pool.coordinate(cCtx)
+	return pool
+}
+
+func (cp *ConnPool) LocalAddrs() []*net.UDPAddr {
+	retval := make([]*net.UDPAddr, len(cp.items))
+	i := 0
+	for conn := range cp.items {
+		retval[i] = conn.LocalAddr().(*net.UDPAddr)
+		i++
+	}
+	return retval
+}
+
+func (cp *ConnPool) Exchange(ctx context.Context, client *dns.Client, msg *dns.Msg) (r *dns.Msg, rtt time.Duration, err error) {
+	conn, err := cp.getConnection(ctx)
+	if err != nil {
+		return nil, time.Duration(0), err
+	}
+	defer cp.releaseConnection(conn)
+	return client.ExchangeWithConn(msg, conn)
+}
+
+func (cp *ConnPool) Close() {
+	cp.cancel()
+	for conn := range cp.items {
+		conn.Close()
+	}
+}
+
+func (cp *ConnPool) coordinate(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case client := <-cp.newArrival:
+			heap.Push(&cp.clients, client)
+		case conn := <-cp.finished:
+			cp.items[conn] = false
+		}
+		for conn, inUse := range cp.items {
+			if !inUse && len(cp.clients) > 0 {
+				cp.items[conn] = true
+				client := heap.Pop(&cp.clients).(*waitingClient)
+				client.returnCh <- conn
```

Comment:
I think we can end up locking on this line if the caller's context times out on line 104 before we manage to send an available connection on the waiting client's return chan. Maybe buffering the channel when we create it on line 94 could mitigate that?

Reply:
But even there, this conn would end up being unusable 🤔

Reply:
aah good catch i'll store the ctx.Done for the client and check it here